File: DailyChallenges/CPP/LeetCode_1721.cpp
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        auto left = head;
        int c = 0;
        while (left){
            if (++c == k) break;
            left = left->next;
        }

        auto right = head, node = left->next;
        while (node){
            node = node->next;
            right = right->next;
        }

        swap(left->val, right->val);
        return head;
    }
};

File: DailyChallenges/Scala/LeetCode_1721.scala
object Solution {
    def swapNodes(head: ListNode, k: Int): ListNode = {
        // first pass
        def getLength(n: ListNode, c: Int): Int = if (n == null) c else getLength(n.next, c+1)
        val l = getLength(head,0)
        
        // second pass
        var skth: ListNode = null
        var ekth: ListNode = null
        def getKth(n: ListNode, i: Int): Unit = if ((skth != null & ekth != null) || n == null) return else {
            if (i == k) skth = n
            if (i == l-k+1) ekth = n
            getKth(n.next, i+1)
        }
        getKth(head,1)
        
        // swap
        val sx = skth.x
        skth.x = ekth.x
        ekth.x = sx
        head
    }
}
